### DISCUSSION TITLE
Why i cant use keypad 3x4 in nanoframework?

### DISCUSSION THREAD

**Discussion by sajjadarashhh** [nanoframework/Home#1253] [Q&A]
I use key matrix feature and wana to connect and read user inputs from key pad 3x4
But that not working!!!!
I copy codes from document and connect each c and  r cables to my board equals to pictures in document
Anyone do this?

**Comment by Ellerbach** [comment]
Can you please point out the technical documentation of your keyboard, ideally the schema used. And also the code you're using.
Usual trap with keypad: they have to use a pullup, so did you setup this properly?

**Reply by sajjadarashhh** [reply]
Wait i send you them

**Comment by sajjadarashhh** [comment]
I use it
But i dont know how to link and connect that to esp32 and how can i get a details from that in nanoframework 
![Uploading 16810560473951615140726043704474.jpg…]()

**Comment by sajjadarashhh** [comment]
@Ellerbach i see some documentation for this in key matrix but that not working with it.

**Reply by Ellerbach** [reply]
Wihtout the documentation, hard to say why it's not working!

**Reply by sajjadarashhh** [reply]
https://arduinogetstarted.com/tutorials/arduino-keypad
This documentation

**Reply by Ellerbach** [reply]
Do you have setu the pool up for each pin?

**Reply by sajjadarashhh** [reply]
What do you meaning for pool up?
I connect each pin equal to thats document to the esp board gpio pins like d1 d13 and...

**Reply by Ellerbach** [reply]
To be more precise, you have to use a hardware pull up for all input pins. See the diagram from the readme: https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/KeyMatrix/README.md#connection-using-on-board-gpio

You have resistors that needs to be used as a pull up

**Reply by sajjadarashhh** [reply]
Yes i tried that but thats not working 😑
I dont know why!
Wait i send to you thats picture howni connected

**Reply by Ellerbach** [reply]
There is no picture, it did not upload properly:
![image](https://user-images.githubusercontent.com/8145578/231378935-6ef6dbbe-9bb2-4f97-8d6e-f8eb8fec2e03.png)

**Comment by sajjadarashhh** [comment]
This is a that two image i said 
But idont know why this is not working when i wana use with c1: 13, c2: 27 c3:26
R1: 25, R2: 33, R3: 32, R4: 35
![IMG_20230503_131545 (1)](https://user-images.githubusercontent.com/72276897/235886499-f86f102d-234b-40ff-a6e6-c056f8bca75c.png)
![IMG_20230503_131427](https://user-images.githubusercontent.com/72276897/235886565-005397c0-16fa-4428-895e-af24c0ff6956.png)

**Reply by josesimoes** [reply]
Something wrong with the images you've uploaded... please double check.

**Reply by sajjadarashhh** [reply]
Now check this
@josesimoes

**Reply by Ellerbach** [reply]
Seems like you are missing the pull ups.

**Reply by sajjadarashhh** [reply]
@Ellerbach what is the pull ups?

**Reply by Ellerbach** [reply]
See form this image: https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/KeyMatrix/4x4kb.png

There are resitors, if you don't have them, the input value won't be correct.

See here for the theory: https://en.wikipedia.org/wiki/Pull-up_resistor
